Task3_1/Queue.cs

```diff
@@ -0,0 +1,107 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Task3_1
+{
+    internal class Queue<T> : IQueue<T>
+    {
+        private int _size;
+        private int _maximumCapacity;
+        private T[] _queue;
+
+        public Queue()
+        {
+            _size = 0;
+            _queue = new T[_size];
+            //The maximum capacity its 32 by default for testing purposes only 
+            _maximumCapacity = 32;
+        }
+
+        public Queue(int maximumCapacity)
+        {
+            _size = 0;
+            _queue = new T[_size];
+            _maximumCapacity = maximumCapacity;
+        }
+
+        public Queue(T[] queue)
+        {
+            _queue = queue;
+            _size = _queue.Length;
+            _maximumCapacity = 32;
+        }
+
+        public object Clone()
+        {
+            T[] _tempQueue = new T[_queue.Length];
+
+            for (int i = 0; i < _queue.Length; i++)
+            {
+                _tempQueue[i] = _queue[i];
+            }
+
+            return new Queue<T>(_tempQueue);
+        }
+
+        public void Dequeue()
+        {
+            if( _queue.Length == 0 )
+            {
+                throw new Exception("Cannot dequeue from an empty queue");
+            }
+
```

Comment:
Reallocate memory for each operation is not a good idea )) Allocate memory once ( _queue = new T[maximumCapacity];) Create pointers (fileds) to the head and tail of Queue, and shift appropriate field within the same array